go-controller/pkg/factory/factory.go

```diff
@@ -329,28 +329,52 @@ func getObjectMeta(objType reflect.Type, obj interface{}) (*metav1.ObjectMeta, e
 	return nil, fmt.Errorf("cannot get ObjectMeta from type %v", objType)
 }
 
-func (wf *WatchFactory) addHandler(objType reflect.Type, namespace string, sel labels.Selector, funcs cache.ResourceEventHandler, processExisting func([]interface{})) *Handler {
+func getServiceSpec(objType reflect.Type, obj interface{}) (*kapi.ServiceSpec, error) {
+
+	if service, ok := obj.(*kapi.Service); ok {
+		return &service.Spec, nil
+	}
+
+	return nil, fmt.Errorf("cannot get Spec from type %v", objType)
+}
+
+func (wf *WatchFactory) addHandler(objType reflect.Type, namespace string, sel labels.Selector, matchSel labels.Selector, funcs cache.ResourceEventHandler, processExisting func([]interface{})) *Handler {
 	inf, ok := wf.informers[objType]
 	if !ok {
 		klog.Fatalf("Tried to add handler of unknown object type %v", objType)
 	}
 
 	filterFunc := func(obj interface{}) bool {
-		if namespace == "" && sel == nil {
+		if namespace == "" && sel == nil && matchSel == nil {
 			// Unfiltered handler
 			return true
 		}
+
 		meta, err := getObjectMeta(objType, obj)
 		if err != nil {
 			klog.Errorf("Watch handler filter error: %v", err)
 			return false
 		}
+
 		if namespace != "" && meta.Namespace != namespace {
 			return false
 		}
+		// Matches on an object's meta.Labels
 		if sel != nil && !sel.Matches(labels.Set(meta.Labels)) {
 			return false
 		}
+
+		//Only will work for services
+		spec, err := getServiceSpec(objType, obj)
+		if err != nil {
+			klog.Errorf("Object is not a Service: %v", err)
+		} else {
+			// Matches on a Servic's spec.Selector
+			if matchSel != nil && !matchSel.Matches(labels.Set(spec.Selector)) {
+				return false
+			}
+		}
```

Comment:
should this be wrapped in an`if matchSel != nil` block and have a return after the klog? 

Am I reading this function wrong but will this log an error for every type that isn't a service?

Reply:
Yeah I just realized that, I will change it

Reply:
actually in this comment is the getServiceSpec function required? we could use the objectType in a block around this too

```
if matchSel != nil && objTyp == serviceTytpe {
[cast to get the object]
}
```
or change the `getObjectMeta()` function to return the objects and just use the other cases that don't require the object.

Just spitballing ideas feel free to ignore